### ISSUE TITLE
How do we handle the 1.0 transition for metadata?

### ISSUE THREAD

**Issue by Dazpoet** [KSP-CKAN/NetKAN#1002]
With (modbreaking) KSP 1.0 around the corner it is somewhat important to get a handle on how much metadata updating will be needed in the near future. By knowing which data is affected by manually set `ksp_version` fields we won't have to check through files when someone ask "why isn't mod X updated to 1.0 in CKAN yet" since if it's on the list it means it requires some manual changes. I did some filtering to get a list of mods in NetKAN with a hardcoded 0.90 in them. This is the list I got

```
ActsEW.netkan                      
ADIOS.netkan                       
AdjustableLandingGear.netkan       
AGExt.netkan                       
AlcubierreStandalone.netkan        
AnimatedDecouplers-x64.netkan      
AnimatedDecouplers-x86.netkan      
AntennaRange.netkan                
AutoAction.netkan                  
BDAnimationModules.netkan          
BetterBuoyancy.netkan              
CommunityResourcePack.netkan       
ContractConfigurator-KerbinSpace...
ControlLock.netkan                 
CrossFeedEnabler.netkan            
CrowdSourcedScience.netkan         
CustomBiomes.netkan                
DeadlyReentry.netkan               
IFILS.netkan                       
Impact.netkan                      
InfernalRobotics.netkan
Karbonite.netkan                   
KarbonitePlus.netkan               
KerbalFlightData.netkan            
KerbalFlightIndicators.netkan      
KerbalWind.netkan                  
KOSMOS-SSPP.netkan                 
kribubble.netkan                   
krispeeder1.netkan                 
KSPAPIExtensions.netkan            
MapResourceOverlay.netkan          
MFS.netkan                         
ModuleRCSFX.netkan                 
NearFutureElectrical.netkan        
NebulaEVAHandrails.netkan          
OblivionAerospace.netkan           
PilotAssistant.netkan              
ProceduralDynamics.netkan          
ProceduralFairings.netkan          
RCSBuildAid.netkan                 
RCSLandAid.netkan                  
RealFuels.netkan                   
Regolith.netkan                    
RemoteTech.netkan                  
SDHI-ServiceModuleSystem.netkan    
SmartParts.netkan                  
SmartStage.netkan                  
Snacks.netkan                      
StageRecovery.netkan               
StockBugFixModules.netkan          
StockDragFix.netkan                
StorkDeliverySystem.netkan         
TestFlight.netkan                  
TWR1.netkan                        
UKS.netkan                         
USI-ART.netkan                     
USI-EXP.netkan                     
USI-FTT.netkan                     
USI-SRV.netkan                     
USITools.netkan                    
VerticalPropulsionEmporium.netkan  
ZeroPointInlineFairings-Lite.netkan
ZeroPointInlineFairings.netkan  
```

65 mods in total and not as many as I thought tbh. These will have to be manually updated and I missed anything strictly hardcoded to version numbers <0.90

I also filtered for files having `ksp_version_min` and that list was rather short I made https://github.com/KSP-CKAN/NetKAN/pull/1003 which adds corresponding `ksp_version_max` fields to them. Not sure if this is the right call or not?

CKAN-meta is a whole other story... Maybe a list could be made from filtering files that don't include `"x_generated_by": "netkan"` ? I expect CKAN-meta to be a lot more hellish to deal with but I think (hope) that most core mods are in NetKAN, save of course for stuff like all the Kopernicus and planetbuilding stuff which is a whole story in itself.

Having never seen a version transition in CKAN before, all of the above might be horribly overkill so any discussion on the subject is deeply appreciated.

**Comment by SirDiazo** [comment]
As a mod author with 6 mods in that list (hard coded to KSP version .90), it comes down to the fact that I know my mods are going to break on KSP update as they are all plugins so I want manual control over which version is listed in the CKAN.

IE: Until I get my mod updated, I want the CKAN to not display my mod on the new version of KSP. Then, when I release the updated version of my mod, I update the NetKan file for compatibility with the new version of KSP. In my reading of the docs, this is the best method I could find of keeping the CKAN clear on which version of my mod works with which version of KSP.

**Comment by Dazpoet** [comment]
@SirDiazo just to make sure, this list is in no way a criticism of any kind :) It's just me trying to gauge the amount of mods that will require manual intervention and some of them might have been added by others than the mod authors and therefore require us to check up on them sometime after release just to see if 1.0 compatibility is there or not. I'm hoping the list will shrink rapidly as mod authors start sending PRs in the wake of 1.0 being released.

As the dust settles we'll know which mods might need a check on the forums to see if compatibility is there yet. Some of the mods (KOSMOS-SSPP being an example) also have automatic updates of `ksp_version` compatibility by being on KerbalStuff but still have hardcoded values for one reason or another, meaning their author might not even recognize that a change in the metadata might be required for it to show up on CKAN.

You are correct in that changing the metadata as it progresses is the best way to keep CKAN clear on what mods work with what version of the game. Mods hosted through KerbalStuff do this automagically as the data on KS is updated but mods hosted through GitHub have manual version fields unless they use `$vref`to have it handled by reading KSP-AVC files.

TL;DR: There will be no sprint post 1.0 to just add 1.0 to all mods for the heck of it. This is a list to make sure we eventually check up on all mods with hardcoded version fields.

**Comment by SirDiazo** [comment]
Ah, okay, I misread your first post a bit.

Having said that, there is one potential issues that I have thought of that I'm wondering how CKAN handles.

-Update my mod and release it.
-Update the NetKan for the mod via PR. PR does not get merged yet.
-A CKAN-Meta update runs with the NetKan still having the old version.
-The PR updating the NetKan gets merged.

What happens in this situation?

**Comment by Dazpoet** [comment]
The NetKAN bot will always overwrite the most current version of mod-metadata to correspond to that in NetKAN. So if CKAN-meta is updated to your new version before the PR is merged a new CKAN-meta file (for the same version) will be generated and overwrite the previous one once the merge has happened.

**Comment by SirDiazo** [comment]
Alright, everything sounds good to me then.

**Comment by severedsolo** [comment]
Well, I'm on that list. My problem is, that my mod relies on Contract Configurator being updated too. 

Now, as CC is already listed as a dependency, am I right in thinking that CKAN won't offer my mod until the dependencies are ready for 1.0?

**Comment by Dazpoet** [comment]
Yes a mod won't offer itself up for download until its dependencies are also available for download.

**Comment by Dazpoet** [comment]
Generated a new list

```
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\ActiveTextureManagement-x64-Aggressive.netkan                                                                                               
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\ActiveTextureManagement-x64-Basic.netkan                                                                                                    
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\ADIOS.netkan                                                                                                                                
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\AnimatedDecouplers-x64.netkan                                                                                                               
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\CustomBiomes.netkan                                                                                                                         
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\FShangarExtender.netkan                                                                                                                     
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\GCMonitor.netkan                                                                                                                            
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\KOSMOS-SSPP.netkan                                                                                                                          
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\kribubble.netkan                                                                                                                            
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\krispeeder1.netkan                                                                                                                          
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\MapResourceOverlay.netkan                                                                                                                   
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\NearFutureElectrical.netkan                                                                                                                 
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\NebulaEVAHandrails.netkan                                                                                                                   
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\OPTMk3MiniExpansion.netkan                                                                                                                  
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\ProceduralDynamics.netkan                                                                                                                   
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\RCSBuildAid.netkan                                                                                                                          
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\RealFuels.netkan                                                                                                                            
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\Regolith.netkan                                                                                                                             
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\SDHI-ServiceModuleSystem.netkan                                                                                                             
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\Snacks.netkan                                                                                                                               
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\StockDragFix.netkan                                                                                                                         
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\USI-ART.netkan                                                                                                                              
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\VerticalPropulsionEmporium.netkan                                                                                                           
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\ZeroPointInlineFairings-Lite.netkan                                                                                                         
C:\Users\Daz\Documents\GitHub\NetKAN\NetKAN\ZeroPointInlineFairings.netkan
```

Seems to be me we've gotten most of them fixed since many mods on this list are supposed to be locked to 0.90 for various reasons.

**Comment by Dazpoet** [comment]
I'm pretty darn happy with how metadata is currently switching over to 1.0.x and as such I'm closing this :)
